packages/lodestar-beacon-state-transition/src/fast/util/interface.ts

```diff
@@ -12,3 +15,128 @@ export type ReadonlyEpochContext = {
   readonly previousShuffling?: IReadonlyEpochShuffling;
   getBeaconProposer: (slot: Slot) => ValidatorIndex;
 };
+
+/**
+ * Cache validators of state using a persistent vector to improve the loop performance.
+ * Instead of accessing `validators` array directly inside BeaconState, use:
+ * + flatValidators() for the loop
+ * + setValidator() for an update
+ * + addValidator() for a creation
+ * that'd update both the cached validators and the one in the original state.
+ */
+// eslint-disable-next-line @typescript-eslint/naming-convention
+export interface CachedValidatorsBeaconState extends BeaconState {
+  flatValidators(): Vector<IFlatValidator>;
+  setValidator(i: ValidatorIndex, value: Partial<IFlatValidator>): void;
+  addValidator(validator: Validator): void;
+  getOriginalState(): BeaconState;
+  clone(): CachedValidatorsBeaconState;
+}
+
+/**
+ * Looping through validators inside TreeBacked<BeaconState> is so expensive.
+ * Cache validators from TreeBacked<BeaconState> using a persistent vector for efficiency.
+ * When write, write to both the cache and TreeBacked.
+ * When read, just return the cache.
+ */
+export class CachedValidatorsBeaconState {
+  // this is for the proxy handler, not for the external use
+  public _state: BeaconState;
+  // this is immutable and shared across BeaconStates for most of the validators
+  private _cachedValidators: Vector<IFlatValidator>;
+
+  constructor(state: BeaconState, cachedValidators: Vector<IFlatValidator>) {
+    this._state = state;
+    this._cachedValidators = cachedValidators;
+  }
+
+  public createProxy(): CachedValidatorsBeaconState {
+    return new Proxy(this, new CachedValidatorsBeaconStateProxyHandler());
+  }
+
+  /**
+   * Write to both the cached validator and BeaconState.
+   * _cachedValidators refers to a new instance
+   */
+  public setValidator(i: ValidatorIndex, value: Partial<IFlatValidator>): void {
+    if (this._cachedValidators) {
+      const validator = this._cachedValidators.get(i);
+      this._cachedValidators = this._cachedValidators.set(i, {...validator!, ...value});
+    }
+    const validator = this._state.validators[i];
+    if (value.activationEligibilityEpoch !== undefined)
+      validator.activationEligibilityEpoch = value.activationEligibilityEpoch;
+    if (value.activationEpoch !== undefined) validator.activationEpoch = value.activationEpoch;
+    if (value.effectiveBalance !== undefined) validator.effectiveBalance = value.effectiveBalance;
+    if (value.exitEpoch !== undefined) validator.exitEpoch = value.exitEpoch;
+    if (value.slashed !== undefined) validator.slashed = value.slashed;
+    if (value.withdrawableEpoch !== undefined) validator.withdrawableEpoch = value.withdrawableEpoch;
```

Comment:
Can't we replace all those if-elses with `Object.assign(validator, value)`. This won't create new instance and it will automatically update it in array? 

It also reduces chance that we forget to update this in case some new prop comes along

Reply:
thanks that's a good ideas, it works (I thought it doesn't initially so I have to check each property), unit tests passed.